libbpf-tools/ttysnoop.c

```diff
@@ -0,0 +1,225 @@
+		}
+		read_bytes = num;
+		break;
+	case 'c':
+		errno = 0;
+		num = strtol(arg, NULL, 10);
+		if (errno || num <= 0) {
+			warn("Invalid data count: %s\n", arg);
+			argp_usage(state);
+		}
+		read_count = num;
+		break;
+	case 'h':
+		argp_state_help(state, stderr, ARGP_HELP_STD_HELP);
+		break;
+	case ARGP_KEY_ARG:
+		if (pts != -1) {
+			warn("Too many devices\n");
+			argp_usage(state);
+		}
+		if (arg[0] != '/')
+			snprintf(path, sizeof(path), "/dev/pts/%s", arg);
+		else
+			strcpy(path, arg);
+		if (stat(path, &statbuf) < 0) {
+			warn("Wrong device\n");
+			argp_usage(state);
+		}
+		pts = statbuf.st_ino;
+		break;
+	case ARGP_KEY_END:
+		if (pts == -1) {
+			warn("Device required\n");
+			argp_usage(state);
+		}
+		break;
+	default:
+		return ARGP_ERR_UNKNOWN;
+	}
+	return 0;
+}
+
+static void sig_int(int signo)
+{
+	exiting = 1;
+}
+
+static void handle_event(void *ctx, int cpu, void *data, __u32 data_sz)
+{
+	struct event *e = data;
+
+	printf("%s", e->buf);
+}
+
+static void handle_lost_events(void *ctx, int cpu, __u64 lost_cnt)
+{
+	warn("lost %llu events on CPU #%d\n", lost_cnt, cpu);
+}
+
+int main(int argc, char **argv)
+{
+	static const struct argp argp = {
+		.options = opts,
+		.parser = parse_arg,
+		.doc = argp_program_doc,
+	};
+	struct perf_buffer_opts pb_opts;
+	struct perf_buffer *pb = NULL;
+	struct ttysnoop_bpf *obj;
+	int err;
+
+	err = argp_parse(&argp, argc, argv, 0, NULL, NULL);
+	if (err)
+		return err;
+
+	err = bump_memlock_rlimit();
+	if (err) {
+		warn("failed to increase rlimit: %d\n", err);
+		return 1;
+	}
+
+	obj = ttysnoop_bpf__open();
+	if (!obj) {
+		warn("failed to open BPF object\n");
+		return 1;
+	}
+
+	obj->rodata->pts = pts;
+	obj->rodata->read_bytes = read_bytes >= BUF_SIZE ? (BUF_SIZE - 1) : read_bytes;
+	obj->rodata->read_count = read_count;
+
+	if (fentry_exists("tty_write", NULL))
+		bpf_program__set_autoload(obj->progs.tty_write_entry, false);
+	else
+		bpf_program__set_autoload(obj->progs.tty_write_fentry, false);
+
+	err = ttysnoop_bpf__load(obj);
+	if (err) {
+		warn("failed to load BPF object: %d\n", err);
+		goto cleanup;
+	}
+
+	err = ttysnoop_bpf__attach(obj);
+	if (err) {
+		warn("failed to attach BPF programs: %d\n", err);
+		goto cleanup;
+	}
+
+	pb_opts.sample_cb = handle_event;
+	pb_opts.lost_cb = handle_lost_events;
+	pb = perf_buffer__new(bpf_map__fd(obj->maps.events), PERF_BUFFER_PAGES, &pb_opts);
+	err = libbpf_get_error(pb);
+	if (err) {
+		warn("failed to open perf buffer: %d\n", err);
+		goto cleanup;
+	}
+
+	if (signal(SIGINT, sig_int) == SIG_ERR) {
+		warn("can't set signal handler: %s\n", strerror(-errno));
+		goto cleanup;
+	}
+
+	if (clear_screen) {
+		err = system("clear");
+		if (err)
+			goto cleanup;
+	}
+
+	while (1) {
+		err = perf_buffer__poll(pb, PERF_POLL_TIMEOUT_MS);
+		if (err == -EINTR) {
+ 			err = 0;
+ 			goto cleanup;
+ 		}
+
+		if (err < 0)
+			break;
+		if (exiting)
+			goto cleanup;
+	}
+	warn("error polling perf buffer: %d\n", err);
```

Comment:
Would you mind following the same structure as in https://github.com/iovisor/bcc/pull/3630 to keep all the tools uniform?